external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,446 @@
+
+    val kafkaParamsForStrategy =
+      ConfigUpdater("source", specifiedKafkaParams)
+        .set(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, deserClassName)
+        .set(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, deserClassName)
+        .build()
+
+    val kafkaParamsForExecutors =
+      ConfigUpdater("source", specifiedKafkaParams)
+        .set(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, deserClassName)
+        .set(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, deserClassName)
+
+        // So that consumers in executors never throw NoOffsetForPartitionException
+        .set(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "none")
+
+        // So that consumers in executors do not mess with user-specified group id
+        .set(ConsumerConfig.GROUP_ID_CONFIG,
+          "spark-executor-" + specifiedKafkaParams(ConsumerConfig.GROUP_ID_CONFIG))
+
+        // So that consumers in executors no keep committing offsets unnecessaribly
+        .set(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
+
+        // If buffer config is not set, it to reasonable value to work around
+        // buffer issues (see KAFKA-3135)
+        .setIfUnset(ConsumerConfig.RECEIVE_BUFFER_CONFIG, 65536: java.lang.Integer)
+        .build()
+
+    val strategy = caseInsensitiveParams.find(x => strategyOptionNames.contains(x._1)).get match {
+      case ("subscribe", value) =>
+        SubscribeStrategy[Array[Byte], Array[Byte]](
+          value.split(",").map(_.trim()).filter(_.nonEmpty),
+          kafkaParamsForStrategy)
+      case ("subscribepattern", value) =>
+        SubscribePatternStrategy[Array[Byte], Array[Byte]](
+          value.trim(),
+          kafkaParamsForStrategy)
+      case _ =>
+        // Should never reach here as we are already matching on
+        // matched strategy names
+        throw new IllegalArgumentException("Unknown option")
+    }
+
+    new KafkaSource(sqlContext, strategy, kafkaParamsForExecutors, parameters)
+  }
+
+  private def validateOptions(parameters: Map[String, String]): Unit = {
+    val caseInsensitiveParams = parameters.map { case (k, v) => (k.toLowerCase, v) }
+    val specifiedStrategies =
+      caseInsensitiveParams.filter { case(k, _) => strategyOptionNames.contains(k) }.toSeq
+    if (specifiedStrategies.isEmpty) {
+      throw new IllegalArgumentException(
+        "One of the following options must be specified for Kafka source: "
+          + strategyOptionNames.mkString(", ") + ". See docs for more details.")
+    } else if (specifiedStrategies.size > 1) {
+      throw new IllegalArgumentException(
+        "Only one of the following options can be specified for Kafka source: "
+          + strategyOptionNames.mkString(", ") + ". See docs for more details.")
+    }
+
+    val strategy = caseInsensitiveParams.find(x => strategyOptionNames.contains(x._1)).get match {
+      case ("subscribe", value) =>
+        val topics = value.split(",").map(_.trim).filter(_.nonEmpty)
```

Comment:
It's not legal to create topics with a comma, verified in code, tested with kafka 0.8 and 0.10